pkg/discovery/dtofactory/container_dto_builder.go

```diff
@@ -0,0 +1,162 @@
+package dtofactory
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	api "k8s.io/client-go/pkg/api/v1"
+
+	"github.com/turbonomic/kubeturbo/pkg/discovery/metrics"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/task"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/util"
+	sdkbuilder "github.com/turbonomic/turbo-go-sdk/pkg/builder"
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+)
+
+var (
+	commoditySold = []metrics.ResourceType{
+		metrics.CPU,
+		metrics.Memory,
+	}
+
+	commodityBought = []metrics.ResourceType{
+		metrics.CPU,
+		metrics.Memory,
+	}
+)
+
+type containerDTOBuilder struct {
+	generalBuilder
+}
+
+func NewContainerDTOBuilder(sink *metrics.EntityMetricSink) *containerDTOBuilder {
+	return &containerDTOBuilder{
+		generalBuilder: newGeneralBuilder(sink),
+	}
+}
+
+// get cpu frequency
+func (builder *containerDTOBuilder) getNodeCPUFrequency(pod *api.Pod) (float64, error) {
+	key := util.NodeKeyFromPodFunc(pod)
+	cpuFrequencyUID := metrics.GenerateEntityStateMetricUID(task.NodeType, key, metrics.CpuFrequency)
+	cpuFrequencyMetric, err := builder.metricsSink.GetMetric(cpuFrequencyUID)
+	if err != nil {
+		err := fmt.Errorf("Failed to get cpu frequency from sink for node %s: %v", key, err)
+		glog.Error(err)
+		return 0.0, err
+	}
+
+	cpuFrequency := cpuFrequencyMetric.GetValue().(float64)
+	return cpuFrequency, nil
+}
+
+func (builder *containerDTOBuilder) BuildDTOs(pods []*api.Pod) ([]*proto.EntityDTO, error) {
+	var result []*proto.EntityDTO
+
+	for _, pod := range pods {
+		nodeCPUFrequency, err := builder.getNodeCPUFrequency(pod)
+		if err != nil {
+			glog.Errorf("failed to build ContainerDTOs for pod[%s]: %v", pod.Name, err)
+			continue
+		}
+		podId := string(pod.UID)
+
+		for i := range pod.Spec.Containers {
+			container := &(pod.Spec.Containers[i])
+
+			containerId := util.ContainerIdFunc(podId, i)
+			name := util.ContainerNameFunc(pod, container)
+			ebuilder := sdkbuilder.NewEntityDTOBuilder(proto.EntityDTO_CONTAINER, containerId).
+				DisplayName(name)
+
+			//1. commodities sold
+			commoditiesSold, err := builder.getCommoditiesSold(containerId, nodeCPUFrequency)
+			if err != nil {
+				glog.Errorf("failed to create commoditiesSold for container[%s]: %v", name, err)
+				continue
+			}
+			ebuilder.SellsCommodities(commoditiesSold)
+
+			//2. commodities bought
+			commoditiesBought, err := builder.getCommoditiesBought(podId, containerId, nodeCPUFrequency)
+			if err != nil {
+				glog.Errorf("failed to create commoditiesBought for container[%s]: %v", name, err)
+				continue
+			}
+			provider := sdkbuilder.CreateProvider(proto.EntityDTO_CONTAINER_POD, podId)
+			ebuilder.Provider(provider).BuysCommodities(commoditiesBought)
+
+			//3. set properties
+			if !util.Monitored(pod) {
```

Comment:
We may want to review this with M2 team. The reason why we set monitored false for some pod is because we don't want our market to MOVE them. But we might still want to be able to resize the container inside the pod. Also now there're new flags (moveable, resizable, etc.), we might not need to use Monitored to disable move of pod. Please ask Wei to confirm.

Reply:
Currently, we cannot resize a static pod, or a pod controlled by DaemonSets yet.  They are not scheduled as the other kinds of pods.